pegasus/test/test_email_preference_helpers.rb

```diff
@@ -0,0 +1,140 @@
+      assert_nil row_id
+    else
+      assert false, "Expected an Opt In validation error."
+    end
+  end
+
+  def test_upsert_email_preference_with_invalid_source_does_not_create_email_preference
+    begin
+      row_id = EmailPreferenceHelper.upsert(
+        email: 'test_valid@example.net',
+        opt_in: true,
+        ip_address: '1.1.1.1',
+        source: 'Where have all the cowboys gone?',
+        form_kind: nil
+      )
+    rescue StandardError => error
+      assert_equal 'Source is not included in the list', error.message
+      assert_nil row_id
+    else
+      assert false, "Expected a source validation error."
+    end
+  end
+
+  def test_upsert_email_preference_without_ip_address_does_not_create_email_preference
+    begin
+      row_id = EmailPreferenceHelper.upsert(
+        email: 'test_valid@example.net',
+        opt_in: true,
+        ip_address: nil,
+        source: EmailPreferenceConstants::ACCOUNT_SIGN_UP,
+        form_kind: nil
+      )
+    rescue StandardError => error
+      assert_equal 'IP Address is required', error.message
+      assert_nil row_id
+    else
+      assert false, "Expected an IP Address error."
+    end
+  end
+
+  def test_upsert_existing_email_preference_changes_existing_email_preference
+    existing_row_id = EmailPreferenceHelper.upsert(
+      email: 'existing@example.net',
+      opt_in: false,
+      ip_address: '1.1.1.1',
+      source: EmailPreferenceConstants::ACCOUNT_SIGN_UP,
+      form_kind: nil
+    )
+
+    updated_row_id = EmailPreferenceHelper.upsert(
+      email: 'existing@example.net',
+      opt_in: true,
+      ip_address: '2.2.2.2',
+      source: EmailPreferenceConstants::FORM_ACCESS_REPORT,
+      form_kind: "0"
+    )
+    updated_email_preference = Dashboard.db[:email_preferences].where(id: updated_row_id).first
+
+    assert_equal updated_row_id, existing_row_id
+    assert_equal true, updated_email_preference[:opt_in]
+    assert_equal '2.2.2.2', updated_email_preference[:ip_address]
+    assert_equal EmailPreferenceConstants::FORM_ACCESS_REPORT, updated_email_preference[:source]
+    assert_equal "0", updated_email_preference[:form_kind]
+    refute_equal updated_email_preference[:updated_at], updated_email_preference[:created_at]
+  end
+
+  def test_upsert_email_preference_that_is_already_opted_in_does_not_opt_out
+    EmailPreferenceHelper.upsert(
+      email: 'opt_in@example.net',
+      opt_in: true,
+      ip_address: '1.1.1.1',
+      source: EmailPreferenceConstants::ACCOUNT_SIGN_UP,
+      form_kind: nil
+    )
+
+    still_opted_in_row_id = EmailPreferenceHelper.upsert(
+      email: 'opt_in@example.net',
+      opt_in: false,
+      ip_address: '2.2.2.2',
+      source: EmailPreferenceConstants::FORM_ACCESS_REPORT,
+      form_kind: "0"
+    )
+
+    updated_email_preference = Dashboard.db[:email_preferences].where(id: still_opted_in_row_id).first
+    assert_equal true, updated_email_preference[:opt_in]
```

Comment:
Why not just `assert updated_email_preference[:opt_in]` ?

Reply:
Done!